assets/js/payment-method-extensions/payment-methods/cod/index.js

```diff
@@ -0,0 +1,85 @@
+/**
+ * External dependencies
+ */
+import { registerPaymentMethod } from '@woocommerce/blocks-registry';
+import { __ } from '@wordpress/i18n';
+import { getSetting } from '@woocommerce/settings';
+import { decodeEntities } from '@wordpress/html-entities';
+
+/**
+ * Internal dependencies
+ */
+import { PAYMENT_METHOD_NAME } from './constants';
+
+const settings = getSetting( 'cod_data', {} );
+const defaultLabel = __(
+	'Cash on delivery',
+	'woo-gutenberg-products-block'
+);
+const label = decodeEntities( settings.title ) || defaultLabel;
+
+/**
+ * @typedef {import('@woocommerce/type-defs/registered-payment-method-props').RegisteredPaymentMethodProps} RegisteredPaymentMethodProps
+ */
+
+/**
+ * Content component
+ */
+const Content = () => {
+	return <div>{ decodeEntities( settings.description || '' ) }</div>;
+};
+
+/**
+ * Label component
+ *
+ * @param {*} props Props from payment API.
+ */
+const Label = ( props ) => {
+	const { PaymentMethodLabel } = props.components;
+	return <PaymentMethodLabel text={ label } />;
+};
+
+/**
+ * Determine whether COD is available for this cart/order.
+ *
+ * @param boolean cartNeedsShipping True if the cart contains any physical/shippable products.
+ * @return boolean True if COD payment method should be displayed as a payment option.
+ */
+const canMakePayment = ( { cartNeedsShipping, selectedShippingMethods } ) => {
+	if ( ! settings.enableForVirtual && ! cartNeedsShipping ) {
+		// Store doesn't allow COD for virtual orders AND
+		// order doesn't contain any shippable products.
+		return false;
+	}
+
+	if ( ! settings.enableForShippingMethods ) {
+		// Store does not limit COD to specific shipping methods.
+		return true;
+	}
+
+	// Look for an not-supported shipping method in the user's selected
+	// shipping methods. If one is found, then COD is not allowed.
+	const selectedNotSupported = Object.values( selectedShippingMethods ).some(
+		( shippingMethodId ) => {
+			return ! settings.enableForShippingMethods.includes(
+				shippingMethodId
+			);
+		}
+	);
+	return ! selectedNotSupported;
+};
+
+/**
+ * Cash on Delivery (COD) payment method config object.
+ */
+const cashOnDeliveryPaymentMethod = {
+	name: PAYMENT_METHOD_NAME,
```

Comment:
While working on this I realised that in core/PHP this is referred to as `id` not `name` (it's a "slug" id IMO, e.g. `stripe` `cod` `bacs`). If we can safely name this part of the API/implementations consistent with core I think we should - i.e. make the prop `id`.

Thoughts @nerrad ? This is a pretty minor detail, but could make things clearer for implementors/devs out there.

Reply:
We do [have a `paymentMethodId` prop](https://github.com/woocommerce/woocommerce-gutenberg-products-block/blob/7f6233dae5667beed59b7123c3d52cb9759634a1/assets/js/blocks-registry/payment-methods/payment-method-config.js#L22) and it is [passed through to the server](https://github.com/woocommerce/woocommerce-gutenberg-products-block/blob/7f6233dae5667beed59b7123c3d52cb9759634a1/assets/js/base/context/cart-checkout/checkout/processor/index.js#L194).

We have a `name` vs `id` because the server processing of a payment method might have the same gateway but a different payment method in that gateway. For instance Stripe gateway classes are used to process both Stripe CC and Stripe payment_request. So `paymentMethodId` is used to refer to the payment gateway on server requests, and name is used internally client side to differentiate payment methods in the registration for constructing the dynamic react nodes displaying the payment methods (css `id`, css classnames etc).

You've already raised [an issue around naming consistency](https://github.com/woocommerce/woocommerce-gutenberg-products-block/issues/2442) and this is something that maybe could be visited there, but for now this is a bit tricky because there's already some inconsistency around how terms are interchangeably used within Woo Core and gateways as well. `name` and `paymentMethodId` were chosen as something we could standardize around in the client side registration. I'm open to revisiting, but if we do this it should be discussed and worked out separately. 

I will say that I disagree with this statement:

> This is a pretty minor detail, but could make things clearer for implementors/devs out there.

I'm all for making things clearer, but I don't think this is a minor detail - in this case changing `name` to `id` won't make anything clearer compared to what it currently is imo.

FWIW, some background on when `paymentMethodId` was introduced is in #2170

Reply:
That makes sense - I [added a comment](https://github.com/woocommerce/woocommerce-gutenberg-products-block/issues/2442#issuecomment-658450682) to #2442 about this, I can understand this is tricky to get right and be consistent with Woo core. 

No urgency with this, will no doubt come up as we integrate more gateways and get feedback on the API.